src/hotspot/share/gc/shenandoah/shenandoahMark.inline.hpp

```diff
@@ -99,19 +99,33 @@ void ShenandoahMark::do_task(ShenandoahObjToScanQueue* q, T* cl, ShenandoahLiveD
     // Avoid double-counting objects that are visited twice due to upgrade
     // from final- to strong mark.
     if (task->count_liveness()) {
-      count_liveness(live_data, obj);
+      count_liveness<GENERATION>(live_data, obj, worker_id);
```

Comment:
Experiment with pulling count_liveness into the closure `cl` which already knows the generation type.

Reply:
See comment at line 60 further above. This will be addressed separately as a refactor to follow to avoid complicating this PR with more changes.